tools/document-pii/src/main.rs

```diff
@@ -0,0 +1,696 @@
+                                self.module_path = module_path;
+                                self.current_type = current_type;
+                            }
+                            EnumOrStruct::Struct(itemstruct) => {
+                                let current_type = self.current_type.clone();
+                                let module_path = self.module_path.clone();
+                                self.module_path = use_path.rsplit_once("::").unwrap().0.to_owned();
+                                self.visit_item_struct(&itemstruct.clone());
+                                self.module_path = module_path;
+                                self.current_type = current_type;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<'ast> Visit<'ast> for TypeVisitor {
+    fn visit_item_struct(&mut self, node: &'ast ItemStruct) {
+        self.current_type = node.ident.to_string();
+        // This should stop any infinite recursion
+        if !CURRENT_PATH
+            .lock()
+            .unwrap()
+            .iter()
+            .any(|x| x.qualified_type_name == self.current_type)
+        {
+            for field in node.fields.iter() {
+                self.visit_field(field);
+            }
+        }
+    }
+
+    fn visit_item_enum(&mut self, node: &'ast ItemEnum) {
+        self.current_type = node.ident.to_string();
+        if !CURRENT_PATH
+            .lock()
+            .unwrap()
+            .iter()
+            .any(|x| x.qualified_type_name == self.current_type)
+        {
+            for variant in node.variants.iter() {
+                for field in variant.fields.iter() {
+                    self.visit_field(field);
+                }
+            }
+        }
+    }
+
+    fn visit_field(&mut self, node: &'ast Field) {
+        // Every time we visit a field, we have to append the field to the current_path, it gets
+        // popped in the end of this function. This is done so that we can store the full path
+        // whenever the field matches a correct PII value.
+        CURRENT_PATH.lock().unwrap().push(TypeAndField {
+            qualified_type_name: self.current_type.clone(),
+            field_name: node
+                .clone()
+                .ident
+                .map(|x| x.to_string())
+                .unwrap_or_else(|| "{{Unnamed}}".to_string()),
+        });
+
+        if has_pii_value(node) {
+            PII_TYPES
+                .lock()
+                .unwrap()
+                .insert(CURRENT_PATH.lock().unwrap().clone());
+        }
+
+        self.recursion(node);
+
+        CURRENT_PATH.lock().unwrap().pop();
+    }
+}
+
+/// if you have a field such as "foo: Foo<Bar<Baz>>" this function can take the type of the field
+/// and return a vector of the types like: ["Foo", "Bar", "Baz"].
+fn type_to_string(ty: &Type) -> Vec<String> {
+    fn recursion(ty: &Type, segments: &mut Vec<String>) {
```

Comment:
it was just to make it possible to return a Vec<String> instead of passing in a mutable vector from the outside, i'll change it back since it only adds an extra line in 'TypeVisitor' recursion function